File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
          deque<pair<int, int>> dq;   //store elements of current window
        vector<int> answer;
        
        for(int i=0; i<nums.size(); ++i){
            
            if(!dq.empty() && dq.front().second <= (i-k)){  //remove front element if it goes out of window size //keeping relevant
                dq.pop_front();
            }
            
            while(!dq.empty() && dq.back().first < nums[i]) //maintaing DESC order
            {
                dq.pop_back();
            }
            dq.push_back(make_pair(nums[i], i));    //adding in right place
            
            if(i >= (k -1)){ ///current max for window
                answer.push_back(dq.front().first); //inc max of current window
            }
        }
        
        
        return answer;
    }
};

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        dq = deque()
        ans = []
        i,j = 0,0

        while j < len(nums):
            while dq and dq[-1] < nums[j]:           #calc
                dq.pop()    #removing smaller elements on left
            dq.append(nums[j])

            if j-i+1 < k:   #move j
                j += 1
            elif j-i+1 == k:  #slide window
                ans.append(dq[0])   #queue's front is max for the window
                if nums[i] == dq[0]:
                    dq.popleft()
                i += 1
                j += 1
        return ans


